src/context/TodosContext.tsx

```diff
@@ -0,0 +1,231 @@
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const userId = useAuthContext();
+
+  const deleteTodoFromServer = async (todoId: number) => {
+    setErrors(null);
+    setLoading(prevState => [...prevState, todoId]);
+    try {
+      await deleteTodo(todoId);
+      setTodos(currentTodos => currentTodos.filter(
+        todo => todo.id !== todoId,
+      ));
+    } catch (error) {
+      setErrors(ErrorType.delete);
+    } finally {
+      setLoading(prevState => prevState.filter(
+        loadingId => loadingId !== todoId,
+      ));
+    }
+  };
+
+  const deleteCompletedTodos = async () => {
+    setErrors(null);
+    try {
+      const completedTodos = todos.filter(todo => todo.completed);
+      const queueOfTodosForDelete = completedTodos
+        .map(todo => {
+          return deleteTodoFromServer(todo.id);
+        });
+
+      await Promise.allSettled(queueOfTodosForDelete);
+    } catch (error) {
+      setErrors(ErrorType.delete);
+    }
+  };
+
+  const addTodoToServer = async (newTodo: Omit<Todo, 'id'>) => {
+    setErrors(null);
+    try {
+      setLoading(prevState => [...prevState, 0]);
+      setTempTodo({ id: 0, ...newTodo });
+      const data = await addTodo(newTodo);
+
+      setTodos(currentTodos => [...currentTodos, data]);
+
+      return true;
+    } catch (error) {
+      setErrors(ErrorType.add);
+
+      return false;
+    } finally {
+      setTempTodo(null);
+      setLoading(prevState => prevState.filter(
+        loadingId => loadingId !== 0,
+      ));
+    }
+  };
+
+  const updateTodoOnServer = async (id: number, editParams: Partial<Todo>) => {
+    setErrors(null);
+    setLoading(prevState => [...prevState, id]);
+    try {
+      const data:Todo = await updateTodo(id, editParams);
+
+      setTodos(currentTodos => currentTodos.map(todo => (
+        todo.id === id
+          ? data
+          : todo
+      )));
+    } catch (error) {
+      setErrors(ErrorType.update);
+    } finally {
+      setLoading(prevState => prevState.filter(
+        loadingId => loadingId !== id,
+      ));
+    }
+  };
+
+  const updateAllTodoOnServer = async (status: boolean) => {
+    setErrors(null);
+    try {
+      const activeTodos = todos.filter(todo => !todo.completed);
+
+      if (activeTodos.length) {
+        const queueOfTodosForCompleted = activeTodos
+          .map(todo => (
+            updateTodoOnServer(todo.id, { completed: status })
+          ));
+
+        await Promise.all(queueOfTodosForCompleted);
+      } else {
+        const completedTodos = todos.filter(todo => todo.completed);
+        const queueOfTodosForActivate = completedTodos
+          .map(todo => (
+            updateTodoOnServer(todo.id, { completed: !status })
+          ));
+
+        await Promise.all(queueOfTodosForActivate);
+      }
+    } catch (error) {
+      setErrors(ErrorType.delete);
+    }
+  };
+
+  useEffect(() => {
+    if (userId) {
```

Comment:
Didnt ur AuthContext doesnt prevent from this?

Reply:
`const AuthContext = createContext<number | null>(null);`
because defaultValue can bee null type
I changed to 0, this is ok?
`const AuthContext = createContext<number>(0);`

Reply:
Tbh u are the first person i see trying to do seperate Auth for this. Maybe simpler way will be just check if USER_ID is set in App and if not render component with correct information